### ISSUE TITLE
`wren memory` module is extremely slow (1m49s per query)

### ISSUE THREAD

**Issue by gaogao110** [Canner/WrenAI#2611]
# `wren memory` module is extremely slow (1m49s per query)

## Description

The `wren memory` module (both `fetch` and `recall`) is extremely slow. A simple `wren memory fetch -q "测试"` takes **~1 minute 49 seconds** to complete, making the semantic memory feature nearly unusable for interactive querying.

## Root Cause

Each CLI invocation is a **brand new Python process**, which must re-load the entire ML stack from scratch:

| Phase | Time | % of total |
|-------|------|------------|
| Import `torch` + `transformers` + `sentence-transformers` | ~74s | 52% |
| Create `SentenceTransformer` model + load weights from disk | ~69s | 48% |
| LanceDB import | ~6s | 4% |
| Vector search (once model is ready) | **~0s** | negligible |

**Total: ~149s** (confirmed via function-level instrumentation)

The embedding model (`paraphrase-multilingual-MiniLM-L12-v2`, 384d) is already cached locally at `~/.cache/huggingface/hub/` — disk I/O of loading model weights + Python import overhead of `torch` and `transformers` account for essentially 100% of the wall-clock time. The actual LanceDB vector search is near-instant.

## Environment

- **OS**: Ubuntu 26.04 LTS (x86_64)
- **Python**: 3.14.4
- **wrenai**: 0.13.1
- **wren-core-py**: 0.7.3
- **lancedb**: 0.33.0
- **sentence-transformers**: 5.6.0
- **transformers**: 5.12.1
- **RAM**: 15 GB (14 GB free during test)
- **Memory backend**: `lancedb` (local)
- **Memory stats**: `query_history: 1091 rows`, `schema_items: 6874 rows`
- **Project**: 364 models, data source: doris

## Reproduction Steps

```bash
time wren memory fetch -q "测试"
```

## Observed Performance

| Metric | Value |
|--------|-------|
| real (wall clock) | 1m49s |
| user (CPU) | 21s |
| sys (CPU) | 7s |

Timing breakdown (from instrumented run):

```
_embed_fn (first access):         73.7s   ← torch/transformers import + model init
  ├─ import transformers.utils:   73.7s
  ├─ import sentence_transformers: 73.5s
  └─ import torch.* (scattered):  ~25s
_get_store (lancedb connect):      5.8s
_search_schema (vector search):  ~153.9s  ← includes model loading above
```

## Expected Behavior

A semantic memory query should complete in **under 5–10 seconds** for an interactive CLI experience.

## Suggestions

1. **CLI warm-up / daemon mode**: Since the CLI spawns a new process per invocation, the entire 2-minute overhead is paid every time. A persistent background daemon (`wren memory serve` or similar) that keeps the model loaded in memory would reduce latency to near-zero for subsequent queries.

2. **Lazy-load model at import time, not query time**: Currently the model is loaded on the first `_embed_fn` property access, which happens during the first `fetch`/`recall`. Pre-loading the model during `MemoryStore.__init__` wouldn't help the first query, but at least the user would see progress earlier.

3. **Model reuse across CLI invocations**: Could the model be serialized/deserialized (e.g., via `mmap` or `safetensors` shared memory) so that subsequent CLI calls skip Python re-import of `torch` + `transformers`?

4. **Verbose/debug flag**: Add `--verbose` to `wren memory` commands to show timing breakdowns (model loading, embedding, vector search) so users can diagnose slowness themselves.

5. **Consider lighter embedding backend**: For a CLI tool, loading the full `transformers` + `torch` stack is very heavy. Alternatives like `fastembed` (ONNX-based, much faster to load) or even a simple token-counting TF-IDF fallback could be considered for the CLI use case.

**Comment by gaogao110** [comment]
The results I obtained through AI testing: Can the Wren Memory module optimize the model loading issue?

**Comment by gaogao110** [comment]
@goldmedal

**Comment by goldmedal** [comment]
@gaogao110 , thanks for the detailed report — confirmed, and we found a likely amplifier on our side.

Our memory extra pulls the CUDA build of torch on Linux. It's just ["lancedb>=0.6", "sentence-transformers>=2.2"] with no CPU-only pin, so linux-x86_64 resolves torch + the whole nvidia-* stack (~2.7 GB of wheels, ~6–7 GB installed) instead of ~200 MB. import torch on that build relocates far more shared objects, which matches what you measured: 109s wall vs only 28s CPU — mostly waiting, not computing.

Could you try these three, in order?

1. Confirm the CUDA build, then switch to CPU-only — likely the biggest single win:
```
python -c "import torch; print(torch.__version__)"   # a +cu13 suffix means CUDA build
pip install --force-reinstall torch --index-url https://download.pytorch.org/whl/cpu
time wren memory fetch -q "测试"
```
2. Run it twice in a row and report both times. If the second run is much faster, cold disk reads dominate; if it stays at ~1m49s, there's a fixed wait to find.
3. If it stays slow, try `HF_HUB_OFFLINE=1 time wren memory fetch -q "测试"`. sentence-transformers doesn't pass local_files_only, so even with the model cached, huggingface_hub still makes per-file metadata requests to huggingface.co — if that host is slow from your network, each stalls until the 10s etag timeout (~9–10 files) before falling back to cache.

> Lazy-load model at import time, not query time: Currently the model is loaded on the first _embed_fn property access, which happens during the first fetch/recall. Pre-loading the model during MemoryStore.__init__ wouldn't help the first query, but at least the user would see progress earlier.
>

Two notes on your analysis: suggestion #2 is already the current behavior (MemoryStore.__init__ eagerly calls warm_up()), and your two phases sum to ~143s against 109s wall, so the split between them is overlapping — hence the questions above.

## Workarounds meanwhile
wren memory describe needs no embedding at all, and `WREN_MEMORY_BACKEND=grep` makes recall instant with zero ML deps (it does not cover fetch yet — that's a gap we'll close). However, I think the number of models is huge. The performance won't be good. 

You can also swap the embedding model via `WREN_EMBEDDING_MODEL.` The model must match the index, so reset and rebuild after changing it:

```
export WREN_EMBEDDING_MODEL=sentence-transformers/static-similarity-mrl-multilingual-v1
wren memory reset && wren memory index
time wren memory fetch -q "测试"
```

Two caveats: this only shrinks the weight-loading half of the cost — the torch import is unaffected — and the other multilingual alternatives aren't much smaller than the current default (its size is dominated by the 250k-token multilingual vocab, not the layers). A static-embedding model like the one above is the variant that actually cuts inference to near-zero while keeping Chinese support, so it's the most interesting one to try; retrieval quality will differ, so please compare a few fetch results before keeping it.

----

On our side we'll: skip the model load when no embedding is needed (status/reset/check/list/dump, and fetch when it takes the full-text path), cache the store in wren serve mcp instead of rebuilding it per tool call, resolve the model offline-first when cached, add your --verbose timings, and evaluate an ONNX embedder (your #5) — extras can't pin a package index, so that's the only way to fix the download size for everyone.

**Comment by gaogao110** [comment]
@goldmedal Thank you very much for your reply. After testing, I found that all Wren commands run extremely slowly when executed inside the Python virtual environment .venv, while they perform normally in the global external environment. The wren memory command takes around 20 seconds as shown in the screenshot. Even so, I feel this speed is still relatively slow. Are there any optimization methods available?

<img width="1544" height="1540" alt="Image" src="https://github.com/user-attachments/assets/79ebd5ca-fc88-4c34-a8db-333aca146393" />

**Comment by goldmedal** [comment]
@gaogao110 Thanks for retesting — the `.venv` vs global observation is useful, but I want to pin one thing down before we chase it.

A correction to my previous comment first: I described the code as it stood in **0.13.0**, and two of those statements are already out of date. **0.13.1** (released 2026-07-21) reworked the memory loading path:

- The embedding model is **no longer loaded in `MemoryStore.__init__`** — it loads lazily, on the first call that actually needs an embedding. So commands and tool calls that need none no longer pay for it. Your suggestion #2 is effectively implemented.
- The model is now resolved **local-first** (`local_files_only=True` first, online load only as a fallback), so an already-cached model no longer waits on huggingface.co metadata requests. The `HF_HUB_OFFLINE=1` workaround I suggested only matters on 0.13.0 and older.
- The loaded model is **cached process-wide**. Within one long-lived `wren serve mcp` process, only the first call that needs an embedding pays the load; later calls reuse it. On 0.13.0, every `MemoryStore(...)` loaded its own copy.

So, before anything else — which version are you running, in each of the two environments?

```
wren --version        # or: pip show wrenai
```

If you are on 0.13.0 or older, please upgrade (`pip install -U wrenai`, currently 0.13.2) and re-run the same test. That alone changes the picture, and it decides what we look at next.

**Comment by gaogao110** [comment]
before test version:
 wrenai 0.13.1 in .venv

<img width="1968" height="1066" alt="Image" src="https://github.com/user-attachments/assets/5abfaf0d-53a0-45d8-bad7-580bb1b2d0cc" />

current version：
 wrenai 0.13.2  in global

**Comment by goldmedal** [comment]
Thanks — that settles the version question, and it rules out my own suggestion: 0.13.1 already contains the lazy-load, offline-first and process-wide model-cache changes, so upgrading to 0.13.2 was never going to be the fix here. Please don't spend more time on the version.

One thing in your `pip show` output is a much stronger lead:

```
Location: /mnt/d/dev/code/github/WrenAI/…/.venv/lib/python3.12/site-packages
```

Your `.venv` sits on `/mnt/d`, i.e. a Windows drive seen from WSL rather than the Linux filesystem. Imports from `/mnt/*` cross the Windows filesystem bridge, which is much slower for this exact access pattern — importing `torch` stats and reads thousands of small files plus several large shared objects. That would also explain why *every* `wren` command is slow in `.venv` but fine globally, since even `wren --version` imports from that path.

Rather than guess further, could you give me these three numbers? Each one eliminates a different possibility, and together they should be enough to close this out.

**1. Is `/mnt/d` really a bridged Windows drive, and what does an import cost from there?**

```
mount | grep ' /mnt/d '
time .venv/bin/python -c "import torch"
```

A `9p` or `drvfs` type confirms the bridge. If that `import torch` alone takes tens of seconds, the filesystem is the story and the fix is to put the virtualenv on the Linux filesystem (`~/.venvs/wren`) while leaving your project where it is.

**2. Which torch build is in that `.venv`?**

```
.venv/bin/python -c "import torch; print(torch.__version__)"
```

A `+cu...` suffix means the CUDA build, which on `linux-x86_64` is ~2.7 GB of wheels including the whole `nvidia-*` stack, versus ~200 MB for the CPU build. On a bridged drive the two problems multiply rather than add.

**3. In the trace you posted, is a *later* tool call in the same `wren serve mcp` session still ~20s, or does it drop to near zero?**

On 0.13.1 the embedding model is cached process-wide, so only the first call that needs an embedding should pay the load. If a second `get_context` in the same session is still ~20s, then something is starting a fresh process per call and the cache never applies — that would be a different problem from the two above, and worth knowing about.

For context on where this is heading regardless of your environment: the per-process cost of importing `torch` and loading the model weights is a real floor that we should not have. #2642 adds `--verbose` stage timings so this is measurable instead of inferred, and #2643 tracks a torch-free ONNX embedding backend, which is what removes both the multi-GB install and that import cost.
